lib/spack/spack/binary_distribution.py

```diff
@@ -37,23 +39,349 @@
+            cache_path = self._index_file_cache.cache_path(cache_key)
+            with self._index_file_cache.read_transaction(cache_key):
+                db._read_from_file(cache_path)
+
+            spec_list = db.query_local(installed=False)
+
+            for indexed_spec in spec_list:
+                dag_hash = indexed_spec.dag_hash()
+                full_hash = indexed_spec._full_hash
+
+                if dag_hash not in self._mirrors_for_spec:
+                    self._mirrors_for_spec[dag_hash] = []
+
+                for entry in self._mirrors_for_spec[dag_hash]:
+                    # A binary mirror can only have one spec per DAG hash, so
+                    # if we already have an entry under this DAG hash for this
+                    # mirror url, we may need to replace the spec associated
+                    # with it (but only if it has a different full_hash).
+                    if entry['mirror_url'] == mirror_url:
+                        if full_hash and full_hash != entry['spec']._full_hash:
+                            entry['spec'] = indexed_spec
+                        break
+                else:
+                    self._mirrors_for_spec[dag_hash].append({
+                        "mirror_url": mirror_url,
+                        "spec": indexed_spec,
+                    })
+        finally:
+            shutil.rmtree(tmpdir)
+
+    def get_all_built_specs(self):
+        spec_list = []
+        for dag_hash in self._mirrors_for_spec:
+            # in the absence of further information, all concrete specs
+            # with the same DAG hash are equivalent, so we can just
+            # return the first one in the list.
+            if len(self._mirrors_for_spec[dag_hash]) > 0:
+                spec_list.append(self._mirrors_for_spec[dag_hash][0]['spec'])
+
+        return spec_list
+
+    def find_built_spec(self, spec):
+        """Look in our cache for the built spec corresponding to ``spec``.
+
+        If the spec can be found among the configured binary mirrors, a
+        list is returned that contains the concrete spec and the mirror url
+        of each mirror where it can be found.  Otherwise, ``None`` is
+        returned.
+
+        This method does not trigger reading anything from remote mirrors, but
+        rather just checks if the concrete spec is found within the cache.
+
+        The cache can be updated by calling ``update_local_index_cache()``.
+
+        Args:
+            spec (Spec): Concrete spec to find
+
+        Returns:
+            An list of objects containing the found specs and mirror url where
+                each can be found, e.g.:
+
+                .. code-block:: python
+
+                    [
+                        {
+                            "spec": <concrete-spec>,
+                            "mirror_url": <mirror-root-url>
+                        }
+                    ]
+        """
+        self.regenerate_spec_cache()
```

Comment:
This looks like what is needed to lazily populate the cache. However, this may end up doing a significant amount of work for each call. I think it would make sense to add a function called `_read()` like:

```
def _read(self):
  if self._mirrors_for_spec:
    return
  self.regenerate_spec_cache()
```

and call that here.

Reply:
In my last round of changes, I added some code to protect against `regenerate_spec_cache()` doing a lot of work if it didn't need to.  There is now a member `_specs_already_associated` which tracks the index hashes for which we've already put the specs into the `_mirrors_for_spec`.  If an index hash is in that dictionary, there should be no need to read the index, iterate the specs, etc, so I avoid it in in that case.  At this point it was my opinion that `rengenerate_spec_cache()` does almost doing nothing if it is called multiple times in succession.  Do you not agree with that assessment?

Reply:
That will work although I think it would be more straightforward if that data structure was a set (and you check if `cached_index_hash in already_associated` vs. `already_associated[cached_index_hash] == True`).

I think I would still have a slight preference for basing this on whether `_mirrors_for_specs` was populated at all, since I think the `regenerate_spec_cache` function will always do the work of populating it (so in some sense it feels slightly redundant to have an additional data structure to track this state).